bin/eslint-plugin-woocommerce/rules/feature-flag.js

```diff
@@ -0,0 +1,201 @@
+/**
+ * Traverse up through the chain of parent AST nodes returning the first parent
+ * the predicate returns a truthy value for.
+ *
+ * @param {Object}   sourceNode The AST node to search from.
+ * @param {Function} predicate  A predicate invoked for each parent.
+ *
+ * @return {?Object } The first encountered parent node where the predicate
+ *                    returns a truthy value.
+ */
+function findParent( sourceNode, predicate ) {
+	if ( ! sourceNode.parent ) {
+		return;
+	}
+
+	if ( predicate( sourceNode.parent ) ) {
+		return sourceNode.parent;
+	}
+
+	return findParent( sourceNode.parent, predicate );
+}
+
+/**
+ * Tests whether the WOOCOMMERCE_BLOCKS_PHASE variable is accessed via
+ * `process.env.WOOCOMMERCE_BLOCKS_PHASE`.
+ *
+ * @example
+ * ```js
+ * // good
+ * if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {
+ *
+ * // bad
+ * if ( WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {
+ * ```
+ *
+ * @param {Object} node    The WOOCOMMERCE_BLOCKS_PHASE identifier node.
+ * @param {Object} context The eslint context object.
+ */
+function testIsAccessedViaProcessEnv( node, context ) {
+	let parent = node.parent;
+
+	if (
+		parent &&
+		parent.type === 'MemberExpression' &&
+		context.getSource( parent ) === 'process.env.WOOCOMMERCE_BLOCKS_PHASE'
+	) {
+		return;
+	}
+	if ( parent && parent.type === 'BinaryExpression' ) {
+		if ( parent.left.type === 'Identifier' ) {
+			parent = parent.left;
+		} else {
+			parent = parent.right;
+		}
+	}
+	context.report( {
+		node,
+		message:
+			'The `WOOCOMMERCE_BLOCKS_PHASE` constant should be accessed using `process.env.WOOCOMMERCE_BLOCKS_PHASE`.',
+		fix( fixer ) {
+			return fixer.replaceText(
+				parent,
+				'process.env.WOOCOMMERCE_BLOCKS_PHASE'
+			);
+		},
+	} );
+}
+
+/**
+ * Tests whether the WOOCOMMERCE_BLOCKS_PHASE variable is used in a strict binary
+ * equality expression in a comparison with a enum of string flags, triggering a
+ * violation if not.
+ *
+ * @example
+ * ```js
+ * // good
+ * if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {
+ * if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'stable' ) {
+ *
+ * // bad
+ * if ( process.env.WOOCOMMERCE_BLOCKS_PHASE == 'experimental' ) {
+ * if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'core' ) {
+ * ```
+ *
+ * @param {Object} node    The WOOCOMMERCE_BLOCKS_PHASE identifier node.
+ * @param {Object} context The eslint context object.
+ */
+function testIsUsedInStrictBinaryExpression( node, context ) {
+	const parent = findParent(
+		node,
+		( candidate ) => candidate.type === 'BinaryExpression'
+	);
+	const flags = [ 'experimental', 'stable' ];
+	let providedFlag;
+	if ( parent ) {
+		const comparisonNode =
+			node.parent.type === 'MemberExpression' ? node.parent : node;
+
+		// Test for process.env.WOOCOMMERCE_BLOCKS_PHASE === <string> or <string> === process.env.WOOCOMMERCE_BLOCKS_PHASE
+		const hasCorrectOperator = [ '===', '!==' ].includes( parent.operator );
```

Comment:
should we only be allowing `===` and not `!==`?

Reply:
Can’t think of a situation in which `!==` will be needed that `stable` can't fill, so we either forbid `stable` or forbid `!==`.
Forbidding `!==` will make it more readable

Reply:
Can’t think of a situation in which `!==` will be needed that `stable` can't fill, so we either forbid `stable` or forbid `!==`.
Forbidding `!==` will make it more readable

Reply:
I apologize but I'm not sure I understand what you're meaning here.  My assumption is that this is a test for whether the operator is `===` or `!==` and I'm suggesting that only `===` should be acceptable. What use-case can you think of where we'd have to do something like the following?

```js
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE !== 'experimental' ) {
    // do something
}
```

Reply:
On a re-reading your comment, I _think_ you are agreeing with my suggestion?

Reply:
yeah I suggested that we can either forbid using `stable`, this makes using `!==` since sometimes you might want to ship legacy code that only exists in stable.
But I prefer we keep stable, so we should remove `!==`, **so yeah I'm agreeing**